### PR TITLE
Assorted minor changes

### PR DESCRIPTION
**Pull request by notnoop**
Noticed few things while reviewing code - each commit is independent.

### CODE REVIEW

File: command/agent/config_parse_test.go [hunk 1]
```diff
@@ -648,24 +650,90 @@ func TestConfig_ParseDir(t *testing.T) {
...
+	require.Empty(t, c.Client.Options)
+	c.Client.Options = nil
+	require.Empty(t, c.Client.Meta)
+	c.Client.Meta = nil
+	require.Empty(t, c.Client.ChrootEnv)
+	c.Client.ChrootEnv = nil
+	require.Empty(t, c.Server.StartJoin)
+	c.Server.StartJoin = nil
+	require.Empty(t, c.HTTPAPIResponseHeaders)
+	c.HTTPAPIResponseHeaders = nil
```
Comment (notnoop) [inline command/agent/config_parse_test.go]: Ideally, we don't modify the package variable as they might be shared across tests.

File: nomad/state/state_store.go [hunk 2]
```diff
@@ -4217,7 +4217,9 @@ func (s *StateSnapshot) DenormalizeAllocationSlice(allocs []*structs.Allocation)
 }
 
 // DenormalizeAllocationDiffSlice queries the Allocation for each AllocationDiff and merges
-// the updated attributes with the existing Allocation, and attaches the Job provided
+// the updated attributes with the existing Allocation, and attaches the Job provided.
+//
+// This should only be called on terminal alloc, particularly stopped or preempted allocs
 func (s *StateSnapshot) DenormalizeAllocationDiffSlice(allocDiffs []*structs.AllocationDiff) ([]*structs.Allocation, error) {
```
Comment (notnoop) [inline nomad/state/state_store.go]: It's not obvious to me how to distinguish an AllocationDiff that's meant for stopping an allocation vs non-destructive update (e.g. service name change), so cannot add an assertion here.

The distinguishing factor seems to be which field of `ApplyPlanResultsRequest` the diff appears in: e.g. `AllocsStopped` vs `AllocsPreempted` vs `AllocsUpdated`.
The distinguishing factor seems to be the field of `ApplyPlanResultsRequest` and whether the diff is

File: command/agent/config_parse_test.go [hunk 3]
```diff
@@ -648,24 +650,90 @@ func TestConfig_ParseDir(t *testing.T) {
...
+			sort.Strings(dirConfig.Files)
+
+			require.EqualValues(t, dirConfig, config)
+		})
+	}
+
+}
+
+// https://stackoverflow.com/a/30226442
+func permutations(arr []string) [][]string {
```
Comment (langmartin) [inline command/agent/config_parse_test.go]: The test looks great to me, is it worth stashing permutations in well known utility package somewhere?

File: command/agent/config_parse_test.go (hunk 1 above)
Comment (langmartin) [reply command/agent/config_parse_test.go]: sgtm

File: command/agent/config_parse_test.go (hunk 3 above)
Comment (notnoop) [reply command/agent/config_parse_test.go]: We can move it or use a permutation library if we see more need for it.  With lack of generics, it's a bit annoying having a generic function for this :(.

**Pull request merged**